torch/csrc/autograd/engine.cpp

```diff
@@ -1,342 +1,148 @@
-#include <Python.h>
-#include <structmember.h>
+#include "torch/csrc/autograd/engine.h"
 
-#include <vector>
-#include <unordered_map>
-#include <deque>
 #include <set>
+#include <THPP/THPP.h>
 
-#include "THP.h"
+using thpp::Tensor;
 
-PyObject *THPEngineClass = NULL;
+namespace torch { namespace autograd {
 
-// used for topological sort
-using dependencies_type = std::unordered_map<THPFunction *, int>;
-// stores gradient buffers
-using grad_list_type = std::vector<THPObjectPtr>;
-// used for need_copy set (to ensure correct gradient buffering)
-using buffer_set_type = std::set<std::pair<size_t, int>>;
-// gradient buffer - a list of gradient tensors + id
-struct grad_buffer_type: public grad_list_type {
-  template<typename... Args>
-  grad_buffer_type(size_t buffer_id, Args&&... args):
-      grad_list_type(std::forward<Args>(args)...),
-      buffer_id(buffer_id) {};
-  grad_buffer_type(grad_buffer_type &&other):
-      grad_list_type(std::move(other)),
-      buffer_id(other.buffer_id) {};
-  grad_buffer_type& operator=(grad_buffer_type &&other) {
-      grad_list_type::operator=(std::move(other));
-      buffer_id = other.buffer_id;
-      return *this;
-  };
-
-  size_t buffer_id;
-};
-// used for the queue of nodes ready for processing
-using ready_queue_type = std::deque<std::pair<THPFunction *, grad_buffer_type>>;
-
-// Computes graph dependencies (using a super simple topological sort)
-void THPEngine_compute_dependencies(std::vector<THPFunction*> queue,
-    dependencies_type& dependencies, ready_queue_type& ready)
-{
-  std::set<THPFunction *> seen;
+auto Engine::compute_dependencies(function_queue& queue, ready_queue_type& ready) -> dependencies {
+  std::unordered_map<std::shared_ptr<Function>, int> dependencies;
+  std::set<Function*> seen;
   while (queue.size() > 0) {
-    THPFunction *fn = queue.back(); queue.pop_back();
-    for (int i = 0; i < fn->num_inputs; i++) {
-      THPFunction *prev_fn = (THPFunction*)fn->previous_functions[i].get();
-      // We can ignore variables (their backprop is called every time we have
-      // gradient ready).
-      if (THPVariable_Check((PyObject*)prev_fn))
+    auto fn = std::move(queue.back()); queue.pop_back();
+    for (auto& prev_fn_pair : fn->previous_functions) {
+      auto& prev_fn = prev_fn_pair.first;
+      if (!prev_fn)
+        continue;
+      if (dynamic_cast<Variable*>(prev_fn.get()))
```

Comment:
I wrote a little microbenchmark and the overhead of dynamic_cast didn't seem much higher than a virtual function call (8ns vs 6ns).

From a design perspective, I don't really like having Variable extend Function, but it's the simplest thing for now.